### PR TITLE
Update encoding for read/subscribe/write request

### PR DESCRIPTION
**Pull request by yunhanw-google**
#### Problem
update encoding for IM read/subscribe/write request and corresponding client/server

#### Change overview
See above

#### Testing
Update the existing unit tests to cover changes

### CODE REVIEW

File: src/app/MessageDef/ReadRequestMessage.h [hunk 1]
```diff
@@ -71,42 +64,40 @@ class Parser : public StructParser
     /**
      *  @brief Get a TLVReader for the AttributePathIBs. Next() must be called before accessing them.
      *
-     *  @param [in] apAttributePathList    A pointer to an attribute path list parser.
+     *  @param [in] apAttributeRequests    A pointer to an attribute path list parser.
```
Comment (yufengwangca) [inline src/app/MessageDef/ReadRequestMessage.h:67]: Need to update the comment which no more point to attribute path

**Review by yufengwangca — APPROVED** [review]

File: src/app/MessageDef/ReadRequestMessage.h (hunk 1 above)
Comment (yunhanw-google) [reply src/app/MessageDef/ReadRequestMessage.h:67]: it is still an attribute path list parser..although spec name this member as attribute request

**Review by jmartinez-silabs — APPROVED** [review]

**Comment by bzbarsky-apple** [comment]
What are the actual changes being made here?  And is there a reason this is one big mega-PR instead of separate PRs for read, subscribe, write?  Please keep in mind that review time goes about as O(N^2) of PR size unless the PR is just a search-and-replace...

**Comment by yunhanw-google** [comment]
> What are the actual changes being made here? And is there a reason this is one big mega-PR instead of separate PRs for read, subscribe, write? Please keep in mind that review time goes about as O(N^2) of PR size unless the PR is just a search-and-replace...

Agree, read and subscribe have several similar encoding stuff, so i made the change in the same PR...

**Review by andy31415 — APPROVED** [review]

**Review by LuDuda — APPROVED** [review]

**Comment by andy31415** [comment]
@yunhanw-google - merge conflicts

**Comment by bzbarsky-apple** [comment]
> Agree, read and subscribe have several similar encoding stuff, so i made the change in the same PR...

I should note that this did not answer the question I actually asked....

**Pull request merged**